decisions/DR007-Hierarchical-or-compositional-traits-for-specifications.md

```diff
@@ -0,0 +1,600 @@
+    float angle
+  }
+  specification Container {
+    bool isEmpty
+  }
+  specification Container.Shot {
+    int startFrame
+    int endFrame
+  }
+```
+
+If we change the hierarchy to share Sequence, Rig and Panoramic now have
+an unwanted startFrame and endFrame property that might not apply. In
+addition, all sequences have to be containers, or all containers have to
+be sequences.
+
+```mermaid
+classDiagram
+  direction TB
+  Blob <-- Geometry
+  Sequence <-- Blob
+  Geometry <-- AnimationCache
+  Geometry <-- Rig
+  Blob <-- Image
+  Image <-- RenderCache
+  Image <-- Panoramic
+  Sequence <-- Shot
+  Container <-- Sequence : everything has to be a container
+  
+  class Blob {
+    str mimeType
+  }
+  class Sequence {
+    int startFrame
+    int endFrame
+  }
+  class Geometry {
+    str upAxis
+  }
+  class AnimationCache {
+    int size
+  }
+  class Rig {
+    str department
+  }
+  class Image {
+    int channels
+  }
+  class RenderCache {
+    int size
+  }
+  class Panoramic {
+    float angle
+  }
+  class Shot {
+  }
+  class Container {
+    bool isEmpty
+  }
+```
+
+```mermaid
+classDiagram
+  direction TB
+  Blob <-- Geometry
+  Sequence <-- Blob
+  Geometry <-- AnimationCache
+  Geometry <-- Rig
+  Blob <-- Image
+  Image <-- RenderCache
+  Image <-- Panoramic
+  Container <-- Shot
+  Sequence <-- Container : all containers have to be sequences
+  
+  class Blob {
+    str mimeType
+  }
+  class Sequence {
+    int startFrame
+    int endFrame
+  }
+  class Geometry {
+    str upAxis
+  }
+  class AnimationCache {
+    int size
+  }
+  class Rig {
+    str department
+  }
+  class Image {
+    int channels
+  }
+  class RenderCache {
+    int size
+  }
+  class Panoramic {
+    float angle
+  }
+  class Shot {
+
+  }
+  class Container {
+    bool isEmpty
+  }
+```
+
+We could copy the sequence specification but since only one parent is
+possible, it has to be copied. If we don't copy the Sequence class, then
+it becomes multiple inheritance.
+
+```mermaid
+classDiagram
+  direction TB
+  Blob <-- Geometry
+  Geometry <-- Sequence_3
+  Geometry <-- Rig
+  Blob <-- Image
+  Sequence_1 <-- RenderCache
+  Image <-- Sequence_1
+  Image <-- Panoramic
+  Sequence_2 <-- Shot
+  Container <-- Sequence_2
+  Sequence_3 <-- AnimationCache
+  
+  class Blob {
+    str mimeType
+  }
+  class Sequence_1 {
+    int startFrame
+    int endFrame
+  }
+  class Sequence_2 {
+    int startFrame
+    int endFrame
+  }
+  class Sequence_3 {
+    int startFrame
+    int endFrame
+  }
+  
+  class Geometry {
+    str upAxis
+  }
+  class AnimationCache {
+    int size
+  }
+  class Rig {
+    str department
+  }
+  class Image {
+    int channels
+  }
+  class RenderCache {
+    int size
+  }
+  class Panoramic {
+    float angle
+  }
+  class Shot {
+
+  }
+  class Container {
+    bool isEmpty
+  }
+```
+
+#### Option 2
+
+This lets us specify only the properties that apply to the
+specification, without duplication.
+
+```mermaid
```

Comment:
[consider][strong] Is it possible to change the presentation of the `Trait` classes so its more obvious they are conceptually different to the resulting `Specification`s, and to Option 1 - where there are only `Specification`s

Reply:
Added comments to the classes for the first diagrams of Option 1 and Option 2 in 78c87c0